### PR TITLE
GPII-3832: Handle production configuration tests in GPC and AWS

### PR DESCRIPTION
**Pull request by klown**
Run the production configuration tests in a GCP dev cluster:

- the `flowmanager`, `preferences`, and `couch` database services run in the cluster as they normally do.
- a new container, `productiontests` runs as a one-shot job that executes the production configuration tests defined in gpii-universal, using the cluster's `flowmanager`, `preferences`, and `couch` database services.

As such, these tests represent an end-to-end test of the system running in production mode wtihin the cloud.

**NOTE**:  This PR is paired with one for [`gpii-version-updater`](https://github.com/gpii-ops/gpii-version-updater/pull/26).

**NOTE**:  the`productiontests` container uses the gpii-universal docker image that includes changes to:
- [production test code](https://github.com/GPII/universal/pull/814),
- [perfomant validation](https://github.com/GPII/universal/pull/811/) (GPII-4062),
- [removal of the ContextManager](https://github.com/GPII/universal/pull/823/) (GPII-3117).

The tag used in `versions.yml` is from the one calculated here:
https://ci.gpii.net/view/Docker/job/docker-gpii-universal-master-calculate-tag/52/console
(All of the containers that are based on `gpii-universal` use this version).

JIRA: https://issues.gpii.net/browse/GPII-3832

### CODE REVIEW

**Comment by mrtyler** [comment]
LGTM.

It's weird that this task never seems to be called. It doesn't look like it was called in @cindyli's PR that added the task https://github.com/gpii-ops/gpii-infra/pull/29/files. Maybe we had a reason for this, or maybe it was an oversight (maybe Cindy remembers)?

My understanding is that this task won't work once your changes for GPII-3333 are merged, so either way removing this task is the right move.

**Comment by amatas** [comment]
LGTM

**Comment by stepanstipl** [comment]
lgtm

**Comment by klown** [comment]
@mrtyler wrote:

> My understanding is that this task won't work once your changes for GPII-3333 are merged, so either way removing this task is the right move

Two comments, one simple and the other a little more involved:
1. Currently, the command run in the container is `node tests/ProductionConfigTests.js`.  The script file `ProductionConfigTests.js` no longer exists in GPII-3333 due to refactoring and re-use of a number of other already-defined tests.  But, that could be replaced by the latest command: `node tests/production/all-tests.js`.
2. There is a pre-load and post-clear of the Couch database that adds/removes GPII keys and PrefsSafes to use by the production tests.  This is done using direct requests to Couch , e.g., a POST request such as "https://couchdb:5984/gpii" sending a GPII key as the payload.

Regarding 2, it works fine in a Vagrant VM context, but likely needs more in a cloud development environment, such as authorization for these database requests.  Furthermore, as discussed at the 27-Mar teleconference, it's not a good idea to be modifying the database in prd, and probably not even in stg.  My current thought is that the target is only the dev-$USER environment.

**Comment by cindyli** [comment]
@mrtyler was right that the task "wait_for_productionConfigTests_complete" was never called. I guess I was focus on getting "wait_for_gpii_ready" task run and missed including this task. Sorry about that.

**Comment by klown** [comment]
Ready for review @stepanstipl @amatas @cindyli

File: shared/charts/gpii-productiontests/values.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  labels:
+    app: couchdb
+  port: http-couchdb
+
+flowmanager:
+    url: http://flowmanager.gpii.svc.cluster.local:80
+    labels:
+        app: flowmanager
+    svcListenport: http
+    flowmanagerListenPort: 8080
```
Comment (stepanstipl) [inline shared/charts/gpii-productiontests/values.yaml]: This block ^ seems to have extra indentation.

Comment (klown) [reply shared/charts/gpii-productiontests/values.yaml]: Good eye, @stepanstipl !

I'll fix it.

**Comment by klown** [comment]
> Thanks @klown, looks good in general 👍 Couple of questions:
> 
> New universal - have there been any other changes we should be aware of in `20190920182823-c3cc79b` tag?

That is the first version of `universal` that includes the modified code for the tests used here.  Given the way that universal is developed, there are other changes that predate that tag, but are not included in the current tag used in `gpii-infra`'s master branch (`20190809170605-f8485e9`).  The tag used here includes changes from 09-Aug-2019 forward until 20-Sep-2019.

Actually I don't know what image of universal is used for the current deployment (`dev`, `stg`, `prd`).  Am I correct that it the one in the master branch's `versions.yml`, namely, `20190809170605-f8485e9`?

> 
> And I was thinking about the scenario of running these tests: Currently, these are deployed as a job and will run once after the cluster is created (only in dev). I guess you can trigger the test again by running smth. like rake redeploy_module[..].

Yes, I do that all the time -- redeploying the single `productiontests` job is much quicker than shutting down everything (`rake destroy`) and then bringing everything back up.

> 
> However if the test fails, the user won't notice unless he/she looks at the status of the job or logs, right?

Right.

> 
> What I was thinking about is:
> 
>   - Is it valuable to run these tests as part of CI in `stg` too? (I think it is)

I agree.

>    - And in `production`? Is there any potential issue?

There is the issue of changing the database.  The tests add and remove users and prefsSafes to the database so that the flowmanager's retrieval and modification of settings can be tested, but at the same time, not affecting any real users or prefsSafes already present.  That's safe in `dev` and likely okay in `stg`, but it's risky to modify the `production` database in `prd`.  We discussed this during an architecture meeting back in March and decided it was safe to run these tests in dev and stg. There are [notes](https://issues.gpii.net/browse/GPII-3832?focusedCommentId=37813&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-37813) in the JIRA from that meeting, with the last bullet, "We need to figure out how to handle data changes in production and record destruction in a safe way".

>   -  Would it make sense to add some plumbing to be able to run the tests in a similar fashion as current smoke tests (rake test_flowmanager)?
> 

It makes sense from the point of view of avoiding a redeploy.  But it sounds mechanically similar to @the-t-in-rtf 's desire for performance tests of the CBFM's GET and PUT `/settings` endpoint.  It's likely that both sets of tests are useful.

I'll think on it.

**Comment by klown** [comment]
@stepanstipl Regarding running these tests in `stg`, I've [modified the pull request](https://github.com/gpii-ops/gpii-infra/pull/347/commits/7a623ac11bad090ed0681f14134c9de66982e63d) so they can.

**Comment by stepanstipl** [comment]
Thanks @klown 

> Am I correct that it the one in the master branch's versions.yml, namely, 20190809170605-f8485e9?

Yes, that is correct.

> @stepanstipl Regarding running these tests in stg, I've modified the pull request so they can.

OK thanks. I think it's good for now. I would go on and merge this, as long as we get some confirmation that all the changes in universal are ok to be deployed to prod.

I think we can look into adding a step, similar to existing `test_...` steps. This would have the benefit of simple redeploy, but mainly we would actually look at whether the test finished successfully and reflect that in the CI run status. (However, let's do this in separate PR so that we can deploy this one :)).

> it was safe to run these tests in dev and stg.

Dev & `stg` sounds good.

> The tests add and remove users and prefsSafes

Do the tests leave behind new data on every run, or do they overwrite/clean up? I.e. will the database size increase after every run?

**Comment by klown** [comment]
> Do the tests leave behind new data on every run, or do they overwrite/clean up? I.e. will the database size increase after every run?

The test records are deleted from the database when the `flowManager` tests are finished.

Before the `flowManager` tests are run, `users`, `prefsSafes`, and `credentials` are added explicitly to the database.  In addition, `access tokens` are created as a by-product of fetching and storing `user` preferences.  A final stage deletes all of the explicitly added records as well as the extra access tokens.

Technically, the database size depends on how Couch deals with deletion.  Deletion is handled in the tests by setting a record's `_deleted` field to `true`. The number of records decreases, but I don't know if Couch expunges them and storage is recovered.  By way of comparison, `flushtokens` does the same thing -- sets `_deleted` to true.  Does the storage size shrink after the `flushtokens` cronjob runs?

**Comment by stepanstipl** [comment]
> The test records are deleted from the database when the `flowManager` tests are finished.

Thanks @klown, that's great 👍 

> Technically, the database size depends on how Couch deals with deletion. Deletion is handled in the tests by setting a record's _deleted field to true. 

Yes, you're correct. AFAIK CouchDB doesn't really delete documents unless purge is called or compaction happens (the first is not recommended for normal operations, and the second will still keep some metadata around), but that's something we'll eventually have to handle on infrastructure level I believe. You can find more details about each  at https://docs.couchdb.org/en/stable/api/database/misc.html & https://docs.couchdb.org/en/latest/maintenance/compaction.html).

> Does the storage size shrink after the `flushtokens` cronjob runs?

Yes, the compaction happens automatically, and some (although not all) space is reclaimed.

**Comment by klown** [comment]
> AFAIK CouchDB doesn't really delete documents unless purge is called ...

@stepanstipl At one time, there was a code in the tests that did make a `/_purge` request.  But, at the time, the version of CouchDB that we used with `AWS` did not support it at all.  It was removed.

> Yes, the compaction happens automatically, and some (although not all) space is reclaimed.

Good to know, thanks :clap:

**Comment by cindyli** [comment]
> I would go on and merge this, as long as we get some confirmation that all the changes in universal are ok to be deployed to prod.

Looking at [the code change](https://github.com/GPII/universal/compare/f8485e9...8ab7ad23c315c84f9525f3c001bb3b24abea6718) in the universal master after the last production deployment, I personally think they should not affect the cloud. To be safe, we could run QAs first with @klown's dev cluster for the testing purpose, then with the staging and production during the deployment.

Let's clarify the process at tomorrow's arch meeting.

**Comment by klown** [comment]
Given [this revert of the gpii-infra](https://github.com/gpii-ops/gpii-infra/pull/561/), this needs to point to some other version of the universal docker image.

**Comment by klown** [comment]
@stepanstipl,
Since @the-t-in-rtf 's validation work was deployed and merged  (PR #563), this PR is now ready.

I've merged in the latest master and checked in my dev cloud.  All looks good.

**Comment by klown** [comment]
> I'm happy to merge this. Thanks @klown
> 
Thanks @stepanstipl :fireworks:
[BTW,](url) is [PR #26](https://github.com/gpii-ops/gpii-version-updater/pull/26) against `gpii-version-updater` still relevant here?

> I still hope we'll follow this with a `test_` step to actually check for the result of these tests as part of the automated suite. @klown is that something you're happy to look into? Or should the ops team pick up the task?

I'm interested, but can't work on it for a while.  [`GPII-4273`](https://issues.gpii.net/browse/GPII-4273) has priority, and it will eventually involve deployment to `stg` and `prd`, with all the required testing and QA.  If ops wants to take on the 'test_` step(s) in the interim, that's fine with me.

**Comment by klown** [comment]
@stepanstipl @natarajaya I have looked at the `productiontests` container logs and the `istio-proxy-manager` logs in `stg` and all looks as expected.

**Pull request merged**